File: src/molgr/fallback/utils/metals/preparation.py
"""Utility helpers for preparing metal-aware fallback inputs."""

from __future__ import annotations

from collections.abc import Sequence
from typing import List, cast

from openbabel import openbabel as ob
from openbabel import pybel

from molgr.config import MolGRConfig
from molgr.fallback.state import MetalPreparationState
from molgr.fallback.utils import consts, dataclasses
from molgr.fallback.utils.electrons import set_unpaired_electron_count
from molgr.fallback.utils.metal_radical_inference import infer_metal_radical_counts


def _build_metal_states(
    obatom: ob.OBAtom,
    *,
    config: MolGRConfig | None = None,
) -> List[dataclasses.MetalAtomPosition]:
    symbol = ob.GetSymbol(obatom.GetAtomicNum())

    def _default_state() -> dataclasses.MetalAtomPosition:
        return dataclasses.MetalAtomPosition(
            idx=obatom.GetIdx(),
            symbol=symbol,
            element_idx=obatom.GetAtomicNum(),
            valence=0,
            radical_num=0,
            position_x=obatom.GetX(),
            position_y=obatom.GetY(),
            position_z=obatom.GetZ(),
        )

    prior = consts.METAL_VALENCE_AVAILABLE_PRIOR.get(symbol, [])
    minor = consts.METAL_VALENCE_AVAILABLE_MINOR.get(symbol, [])
    seen_valences = set()
    valences: List[int] = []
    for valence in prior + minor:
        if valence in seen_valences:
            continue
        seen_valences.add(valence)
        valences.append(valence)
    if not valences:
        valences = [0]

    if symbol not in consts.METAL_F_D_S_P_ELECTRONS:
        return [_default_state()]

    states: List[dataclasses.MetalAtomPosition] = []
    for valence in valences:
        radicals = infer_metal_radical_counts(obatom, valence, config=config)
        for radical_num in radicals:
            states.append(
                dataclasses.MetalAtomPosition(
                    idx=obatom.GetIdx(),
                    symbol=symbol,
                    element_idx=obatom.GetAtomicNum(),
                    valence=valence,
                    radical_num=radical_num,
                    position_x=obatom.GetX(),
                    position_y=obatom.GetY(),
                    position_z=obatom.GetZ(),
                )
            )

    if not states:
        return [_default_state()]
    return states


def combine_metal_with_omol(
    omol: pybel.Molecule,
    metal_list: Sequence[dataclasses.MetalAtomPosition],
) -> pybel.Molecule:
    """Insert selected metal charge/spin states into the no-metal winner.

    ``metal.radical_num`` is copied only to the real unpaired-electron field.
    Metals receive no active-lone-pair or unresolved-center label; those concepts
    are not inferred during metal-state reinsertion. Organic electron fields are
    preserved by cloning the no-metal molecule.
    """

    obmol = cast(ob.OBMol, omol.clone.OBMol)
    obmol.BeginModify()
    try:
        num_organic = obmol.NumAtoms()
        num_metals = len(metal_list)
        total_atoms = num_organic + num_metals

        for metal in metal_list:
            atom = cast(ob.OBAtom, obmol.NewAtom())
            atom.SetAtomicNum(metal.element_idx)
            atom.SetFormalCharge(metal.valence)
            set_unpaired_electron_count(atom, metal.radical_num)
            atom.SetVector(metal.position_x, metal.position_y, metal.position_z)

        new_order = [0] * total_atoms
        has_error = False
        for i, metal in enumerate(metal_list):
            current_idx = num_organic + 1 + i
            target_slot = metal.idx - 1
            if target_slot < 0 or target_slot >= total_atoms:
                has_error = True
                continue
            if new_order[target_slot] != 0:
                has_error = True
                continue
            new_order[target_slot] = current_idx

        if not has_error:
            current_organic_idx = 1
            for i in range(total_atoms):
                if new_order[i] != 0:
                    continue
                if current_organic_idx > num_organic:
                    has_error = True
                    break
                new_order[i] = current_organic_idx
                current_organic_idx += 1

        if not has_error and all(idx > 0 for idx in new_order):
            obmol.RenumberAtoms(new_order)
    finally:
        obmol.EndModify()
    return pybel.Molecule(obmol)


def prepare_metal_state(
    xyz_block: str,
    total_charge: int,
    total_radical_electrons: int,
    *,
    config: MolGRConfig | None = None,
) -> MetalPreparationState:
    """Split the input into a no-metal XYZ block plus per-metal state options."""

    omol = pybel.readstring("xyz", xyz_block)
    removable_metal_atoms = [
        cast(ob.OBAtom, atom.OBAtom) for atom in omol.atoms if atom.OBAtom.IsMetal()
    ]
    available_valence_radical_states = tuple(
        tuple(_build_metal_states(obatom, config=config)) for obatom in removable_metal_atoms
    )
    if not removable_metal_atoms:
        return MetalPreparationState(
            no_metal_xyz_block=xyz_block,
            available_valence_radical_states=available_valence_radical_states,
            total_charge=total_charge,
            total_radical_electrons=total_radical_electrons,
            phase_history=(
                "read_xyz",
                "build_metal_state_options",
                "preserve_no_metal_xyz",
            ),
            metadata={"metal_atom_count": 0},
        )
    for obatom in removable_metal_atoms:
        omol.OBMol.DeleteAtom(obatom)
    return MetalPreparationState(
        no_metal_xyz_block=cast(str, omol.write("xyz")),
        available_valence_radical_states=available_valence_radical_states,
        total_charge=total_charge,
        total_radical_electrons=total_radical_electrons,
        phase_history=(
            "read_xyz",
            "build_metal_state_options",
            "remove_metal_atoms",
            "serialize_no_metal_xyz",
        ),
        metadata={"metal_atom_count": len(removable_metal_atoms)},
    )


__all__ = [
    "_build_metal_states",
    "combine_metal_with_omol",
    "prepare_metal_state",
]

File: src/cpp/src/utils/metals/preparation.cpp
#include "molgr/utils/metals/preparation.h"

#include "molgr/utils/consts.h"
#include "molgr/utils/electrons.h"
#include "molgr/utils/logger.h"
#include "molgr/utils/metal_radical_inference.h"
#include "molgr/utils/xyz.h"

#include <openbabel/elements.h>
#include "molgr/compat/openbabel_iter.h"

#include <set>

namespace molgr
{
    namespace metal
    {
        namespace preparation
        {
            std::set<int> GetPossibleMetalRadicals(const std::string &metal_symbol, int valence)
            {
                return molgr::GetPossibleMetalRadicals(metal_symbol, valence);
            }

            std::vector<molgr::metal::MetalAtomPosition> BuildMetalStates(
                OpenBabel::OBAtom &obatom,
                const molgr::config::MolGRConfig &config)
            {
                const int atomic_num = static_cast<int>(obatom.GetAtomicNum());
                const std::string symbol = OpenBabel::OBElements::GetSymbol(atomic_num);

                const auto default_state = [&]()
                {
                    return molgr::metal::MetalAtomPosition{
                        static_cast<int>(obatom.GetIdx()),
                        symbol,
                        atomic_num,
                        0,
                        0,
                        obatom.GetX(),
                        obatom.GetY(),
                        obatom.GetZ()};
                };

                std::vector<int> valences;
                std::set<int> seen_valences;
                const auto add_valences = [&](const std::vector<int> &source)
                {
                    for (const int valence : source)
                    {
                        if (seen_valences.insert(valence).second)
                        {
                            valences.push_back(valence);
                        }
                    }
                };

                if (kMetalValencePrior.count(symbol))
                {
                    add_valences(kMetalValencePrior.at(symbol));
                }
                if (kMetalValenceMinor.count(symbol))
                {
                    add_valences(kMetalValenceMinor.at(symbol));
                }
                if (valences.empty())
                {
                    valences.push_back(0);
                }

                if (!kMetalFDSP.count(symbol))
                {
                    return {default_state()};
                }

                std::vector<molgr::metal::MetalAtomPosition> states;
                for (const int valence : valences)
                {
                    const auto radicals =
                        molgr::metal_radical_inference::InferMetalRadicalCounts(
                            obatom,
                            valence,
                            config);
                    for (const int radical_num : radicals)
                    {
                        states.push_back(
                            molgr::metal::MetalAtomPosition{
                                static_cast<int>(obatom.GetIdx()),
                                symbol,
                                atomic_num,
                                valence,
                                radical_num,
                                obatom.GetX(),
                                obatom.GetY(),
                                obatom.GetZ()});
                    }
                }

                if (states.empty())
                {
                    return {default_state()};
                }
                return states;
            }

            // Electron bookkeeping: copy the chosen metal radical_num only into
            // the real unpaired-electron field. Metals receive no active lone-pair
            // or unresolved-center classification; organic fields survive cloning.
            void ReinsertMetalStates(
                OpenBabel::OBMol &mol,
                const std::vector<molgr::metal::MetalAtomPosition> &metals)
            {
                mol.BeginModify();

                int num_organic = mol.NumAtoms();
                int num_metals = static_cast<int>(metals.size());
                int total_atoms = num_organic + num_metals;

                for (const auto &m : metals)
                {
                    OpenBabel::OBAtom *atom = mol.NewAtom();
                    atom->SetAtomicNum(m.element_idx);
                    atom->SetFormalCharge(m.valence);
                    molgr::utils::SetUnpairedElectronCount(*atom, m.radical_num);
                    atom->SetVector(m.position_x, m.position_y, m.position_z);
                }

                std::vector<int> new_order(total_atoms, 0);
                bool error_flag = false;

                for (int i = 0; i < num_metals; ++i)
                {
                    int current_idx = num_organic + 1 + i;
                    int target_slot = metals[i].idx - 1;

                    if (target_slot >= 0 && target_slot < total_atoms)
                    {
                        if (new_order[target_slot] != 0)
                        {
                            LOG_ERROR("[ReinsertMetalStates] Index collision at slot " << target_slot);
                            error_flag = true;
                        }
                        new_order[target_slot] = current_idx;
                    }
                    else
                    {
                        LOG_ERROR("[ReinsertMetalStates] Original index out of bounds: " << metals[i].idx);
                        error_flag = true;
                    }
                }

                int current_organic_idx = 1;
                for (int i = 0; i < total_atoms; ++i)
                {
                    if (new_order[i] == 0)
                    {
                        if (current_organic_idx <= num_organic)
                        {
                            new_order[i] = current_organic_idx;
                            current_organic_idx++;
                        }
                        else
                        {
                            LOG_ERROR("[ReinsertMetalStates] Not enough organic atoms to fill slots.");
                            error_flag = true;
                        }
                    }
                }

                for (int idx : new_order)
                {
                    if (idx == 0)
                    {
                        LOG_ERROR("[ReinsertMetalStates] Invalid 0 index in renumber map. Aborting renumber.");
                        error_flag = true;
                        break;
                    }
                }

                if (!error_flag)
                {
                    mol.RenumberAtoms(new_order);
                }

                mol.EndModify();
            }

            void CombineMetalWithOmol(
                OpenBabel::OBMol &mol,
                const std::vector<molgr::metal::MetalAtomPosition> &metals)
            {
                ReinsertMetalStates(mol, metals);
            }

            molgr::state::MetalPreparationState PrepareMetalState(
                const std::string &xyz_block,
                int total_charge,
                int total_radical_electrons,
                const molgr::config::MolGRConfig &config)
            {
                OpenBabel::OBMol mol;
                if (!molgr::utils::ReadXyzBlockToMol(xyz_block, &mol))
                {
                    return {};
                }

                std::vector<OpenBabel::OBAtom *> removable_metal_atoms;
                std::vector<std::vector<molgr::metal::MetalAtomPosition>> available_states;
                FOR_ATOMS_OF_MOL(atom_iter, mol)
                {
                    OpenBabel::OBAtom *atom = &(*atom_iter);
                    if (!atom->IsMetal())
                    {
                        continue;
                    }
                    removable_metal_atoms.push_back(atom);
                    available_states.push_back(BuildMetalStates(*atom, config));
                }

                for (OpenBabel::OBAtom *atom : removable_metal_atoms)
                {
                    mol.DeleteAtom(atom);
                }

                molgr::state::MetalPreparationState state;
                state.no_metal_xyz_block = molgr::utils::WriteXyzBlock(mol);
                state.available_valence_radical_states = std::move(available_states);
                state.total_charge = total_charge;
                state.total_radical_electrons = total_radical_electrons;
                state.phase_history = {
                    "read_xyz",
                    "build_metal_state_options",
                    "remove_metal_atoms",
                    "serialize_no_metal_xyz",
                };
                state.metadata["metal_atom_count"] = static_cast<int>(removable_metal_atoms.size());
                return state;
            }
        }
    }
}
